File: js/packages/botbuilder-m365/src/DefaultTurnStateManager.ts
/**
 * @module botbuilder-m365
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

// TODO:
/* eslint-disable security/detect-object-injection */
import { TurnContext, Storage, StoreItems } from 'botbuilder';
import { TurnState, TurnStateEntry, TurnStateManager } from './TurnState';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DefaultConversationState {}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DefaultUserState {}

export interface DefaultTempState {
    /**
     * Input passed to an AI prompt
     */
    input: string;

    /**
     * Formatted conversation history for embedding in an AI prompt
     */
    history: string;

    /**
     * Output returned from an AI prompt or function
     */
    output: string;
}
/**
 * Types of Turn State held in MemoryStorage
 *
 * @module botbuilder-m365
 */
export interface DefaultTurnState<
    TConversationState extends DefaultConversationState = DefaultConversationState,
    TUserState extends DefaultUserState = DefaultUserState,
    TTempState extends DefaultTempState = DefaultTempState
> extends TurnState {
    conversation: TurnStateEntry<TConversationState>;
    user: TurnStateEntry<TUserState>;
    temp: TurnStateEntry<TTempState>;
}

export class DefaultTurnStateManager<
    TConversationState extends DefaultConversationState = DefaultConversationState,
    TUserState extends DefaultUserState = DefaultUserState,
    TTempState extends DefaultTempState = DefaultTempState
> implements TurnStateManager<DefaultTurnState<TConversationState, TUserState, TTempState>>
{
    public async loadState(
        storage: Storage,
        context: TurnContext
    ): Promise<DefaultTurnState<TConversationState, TUserState, TTempState>> {
        // Compute state keys
        const activity = context.activity;
        const channelId = activity?.channelId;
        const botId = activity?.recipient?.id;
        const conversationId = activity?.conversation?.id;
        const userId = activity?.from?.id;

        if (!channelId) {
            throw new Error('missing context.activity.channelId');
        }

        if (!botId) {
            throw new Error('missing context.activity.recipient.id');
        }

        if (!conversationId) {
            throw new Error('missing context.activity.conversation.id');
        }

        if (!userId) {
            throw new Error('missing context.activity.from.id');
        }

        const conversationKey = `${channelId}/${botId}/conversations/${conversationId}`;
        const userKey = `${channelId}/${botId}/users/${userId}`;

        // Read items from storage provider (if configured)
        const items = storage ? await storage.read([conversationKey, userKey]) : {};

        // Map items to state object
        const state: DefaultTurnState<TConversationState, TUserState, TTempState> = {
            conversation: new TurnStateEntry(items[conversationKey], conversationKey),
            user: new TurnStateEntry(items[userKey], userKey),
            temp: new TurnStateEntry({} as TTempState)
        };

        return state;
    }

    public async saveState(
        storage: Storage,
        context: TurnContext,
        state: DefaultTurnState<TConversationState, TUserState, TTempState>
    ): Promise<void> {
        // Find changes and deletions
        let changes: StoreItems | undefined;
        let deletions: string[] | undefined;
        for (const key in state) {
            if (Object.prototype.hasOwnProperty.call(state, key)) {
                const entry = state[key];
                if (entry.storageKey) {
                    if (entry.isDeleted) {
                        // Add to deletion list
                        if (deletions) {
                            deletions.push(entry.storageKey);
                        } else {
                            deletions = [entry.storageKey];
                        }
                    } else if (entry.hasChanged) {
                        // Add to change set
                        if (!changes) {
                            changes = {};
                        }

                        changes[entry.storageKey] = entry.value;
                    }
                }
            }
        }

        // Do we have a storage provider?
        if (storage) {
            // Apply changes
            const promises: Promise<void>[] = [];
            if (changes) {
                promises.push(storage.write(changes));
            }

            // Apply deletions
            if (deletions) {
                promises.push(storage.delete(deletions));
            }

            // Wait for completion
            if (promises.length > 0) {
                await Promise.all(promises);
            }
        }
    }
}

File: python/teamsai/turn_state/default_turn_state_manager.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from dataclasses import dataclass
from typing import TypeVar, Generic
from botbuilder.core import TurnContext, Storage

from .turn_state_entry import TurnStateEntry
from .turn_state_manager import TurnState, TurnStateManager


@dataclass
class DefaultConversationState:
    """
    inherit a new interface from this base interface to strongly type
    the applications conversation state
    """


@dataclass
class DefaultUserState:
    """
    inherit a new interface from this base interface to strongly type
    the applications user state
    """


@dataclass
class DefaultTempState:
    """
    inherit a new interface from this base interface to strongly type
    the applications temp state
    """

    input: str
    "input passed to an AI prompt"

    history: str
    "formatted conversation history for embedding in an AI prompt"

    output: str
    "output returned from an AI prompt or function"


ConversationStateT = TypeVar("ConversationStateT", DefaultConversationState)
UserStateT = TypeVar("UserStateT", DefaultUserState)
TempStateT = TypeVar("TempStateT", DefaultTempState)


class DefaultTurnState(TurnState, Generic[ConversationStateT, UserStateT,
                                          TempStateT]):
    """
    defines the default state scopes persisted by the `DefaultTurnStateManager`
    """

    conversation: TurnStateEntry[ConversationStateT]
    user: TurnStateEntry[UserStateT]
    temp: TurnStateEntry[TempStateT]


class DefaultTurnStateManager(
        TurnStateManager[DefaultTurnState[ConversationStateT, UserStateT,
                                          TempStateT]],
        Generic[ConversationStateT, UserStateT, TempStateT]):
    """
    default turn state manager implementation
    """

    async def load_state(
        self, storage: Storage, context: TurnContext
    ) -> DefaultTurnState[ConversationStateT, UserStateT, TempStateT]:
        """
        loads all of the state scopes for the current turn\n
        `storage`: storage provider to load state scopes from\n
        `context`: context for the current turn of conversation with the user
        """

        if not context.activity:
            raise ValueError("missing context.activity")
        if not context.activity.channel_id:
            raise ValueError("missing context.activity.channel_id")
        if not context.activity.recipient:
            raise ValueError("missing context.activity.recipient")
        if not context.activity.recipient.id:
            raise ValueError("missing context.activity.recipient.id")
        if not context.activity.conversation:
            raise ValueError("missing context.activity.conversation")
        if not context.activity.conversation.id:
            raise ValueError("missing context.activity.conversation.id")
        if not context.activity.from_property:
            raise ValueError("missing context.activity.from_property")
        if not context.activity.from_property.id:
            raise ValueError("missing context.activity.from_property.id")

        channel_id = context.activity.channel_id
        convo_id = context.activity.conversation.id
        bot_id = context.activity.recipient.id
        user_id = context.activity.from_property.id
        convo_key = f"{channel_id}/{bot_id}/conversations/{convo_id}"
        user_key = f"{channel_id}/{bot_id}/users/{user_id}"

        items = {}

        if storage:
            items = await storage.read([convo_key, user_key])

        return DefaultTurnState[ConversationStateT, UserStateT, TempStateT](
            conversation=TurnStateEntry(items[convo_key], convo_key),
            user=TurnStateEntry(items[user_key], user_key),
            temp=TurnStateEntry({}))
